### PR TITLE
update to geoarrow 0.3. fixes #56

### LINKED ISSUE #56: Update for latest geoarrow
I'd love to update h3arrow for the latest geoarrow-rs, and potentially use `h3arrow` as an (optional) dependency from [Lonboard](https://github.com/developmentseed/lonboard), a Python library for large geospatial vector data visualization.

Some changes:

- As of this week (unreleased) geoarrow-rs does not have an `O: OffsetSizeTrait` generic on the `PolygonArray`: https://github.com/geoarrow/geoarrow-rs/pull/803. In order to overflow the offsets of an `int32`, you'd need to have more than `2^32` coordinates in a single contiguous array, which would take up 32GB alone. Removing this generic makes it easier to move faster in geoarrow-rs and took out a ton of complexity.
- On the contrary, I think supporting multiple dimensions in geoarrow-rs _is_ necessary complexity. So arrays now have a `const D: usize` generic for the physical dimension. So now when you construct arrays you'd have `PolygonArray<2>`.
- I started to implement a [`pyo3-geoarrow` Rust helper module](https://github.com/geoarrow/geoarrow-rs/tree/main/python/pyo3-geoarrow) to make it easier to handle zero-copy data interop from Python, but that's not yet fully standardized.

Let me know if I can help or give pointers to any of this.

### PR DESCRIPTION
**Pull request by nmandery**

### CODE REVIEW

File: h3ronpy/pyproject.toml [hunk 1]
```diff
@@ -36,7 +36,7 @@ polars = [
     "polars>=1.0"
 ]
 pandas = [
-    "geopandas>=0.10",
+    "geopandas>=1",
```
Comment (kylebarron) [inline h3ronpy/pyproject.toml:39]: 🚀

File: crates/h3arrow/Cargo.toml [hunk 2]
```diff
@@ -21,8 +21,7 @@ spatial_index = ["dep:rstar"]
 [dependencies]
 ahash = "0.8"
 arrow = { workspace = true }
-# geoarrow = { package = "geoarrow", version = "0.1.0", optional = true, features = ["geozero"] }
-geoarrow = { package = "geoarrow", git = "https://github.com/geoarrow/geoarrow-rs.git", rev = "30c18f7c07432dfb92f9612efa54aec1d9392daf", optional = true, features = ["geozero"] }
+geoarrow = { version = "0.3", optional = true, features = ["geozero"] }
```
Comment (kylebarron) [inline crates/h3arrow/Cargo.toml:24]: I just published 0.4.0-beta.1 if you want to bump all the way up to that. You'll just have to take off the `O` parameter.

But note there's also now a separation between "native" GeoArrow arrays and "serialized" GeoArrow arrays. WKB is a "serialized" array.

If you get stuck I'm happy to help update this code

**Review by kylebarron — APPROVED** [review]

File: crates/h3arrow/Cargo.toml (hunk 2 above)
Comment (nmandery) [reply crates/h3arrow/Cargo.toml:24]: I am bit out of the loop on what is happening in the arrow-community. I remember the "native" geoarrow arrays where a bit problematic as there was no way to attach the metadata describing the geometrytype to the array. Is this now possible? Is it feasible to exchange the "native" arrays with other libraries?

Comment (kylebarron) [reply crates/h3arrow/Cargo.toml:24]: No, that's not what I mean. I mean the GeoArrow format has both "Native encoding" and "Serialized encoding" https://geoarrow.org/format.html.

Previously in geoarrow-rs we had a single `GeometryArrayTrait` that all geoarrow array types implemented, but now we have two:

- [`NativeArray`](https://docs.rs/geoarrow/0.4.0-beta.1/geoarrow/trait_/trait.NativeArray.html), which is implemented by `PointArray`, `LineStringArray`, `PolygonArray`, `MultiPointArray`, `MultiLineStringArray`, `MultiPolygonArray`, `MixedGeometryArray`, and `GeometryCollectionArray`.
- [`SerializedArray`](https://docs.rs/geoarrow/0.4.0-beta.1/geoarrow/trait_/trait.SerializedArray.html), which is implemented by `WKBArray` and `WKTArray`.

I just mean that you'll have to update the code to use these new traits

Comment (nmandery) [reply crates/h3arrow/Cargo.toml:24]: Ok, now I got it ;)

BTW: Directly upgrading to geoarrow 0.4.0-beta.1 is not possible for me currently. This requires at least he release of rust-numpy built ontop of pyo3 0.22

Comment (kylebarron) [reply crates/h3arrow/Cargo.toml:24]: I'm also hoping that will come out very soon. https://github.com/PyO3/rust-numpy/pull/442#issuecomment-2384038506 and https://github.com/PyO3/rust-numpy/pull/435 have been merged, which is very promising

**Pull request merged**